### PR TITLE
Add a wrapper for FossID scanner

### PR DESCRIPTION
**Pull request by nnobelis**
This scanner communicates with the FossID API client that
was previously contributed.
In this first implementation, the scanner creates a
project, a scan, runs the scan and fetches the scan
results.

- Add a parameter to FossIdRestService.createScan to be
able to configure scan name before scan's creation.

- Allow a scanner to modify its configuration before
putting the latter in the ScannerRun.

The FossId server configuration is passed through ort.conf :

```
options {
       FossId {
             serverUrl="https://xxx"
             apiKey="xxx"
             user="xxx"
       }
}
```

[1]: https://fossid.com/

Signed-off-by: Nicolas Nobelis <nicolas.nobelis@bosch.io>

### CODE REVIEW

File: scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import org.ossreviewtoolkit.clients.fossid.model.identification.identifiedFiles.IdentifiedFile
+import org.ossreviewtoolkit.clients.fossid.model.identification.ignored.IgnoredFile
+import org.ossreviewtoolkit.clients.fossid.model.identification.markedAsIdentified.MarkedAsIdentifiedFile
+import org.ossreviewtoolkit.clients.fossid.model.summary.Summarizable
+import org.ossreviewtoolkit.model.CopyrightFinding
+import org.ossreviewtoolkit.model.LicenseFinding
+import org.ossreviewtoolkit.model.TextLocation
+
+/**
+ * A simple bean to hold FossId raw results
```
Comment (oheger-bosch) [inline scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt:31]: The term "bean" is incorrect; this class is not compliant to the Java Beans specification. Write "data class" instead. And don't miss the dot at the end of the sentence ;-)

File: scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ */
+internal data class RawResults(
+    val identifiedFiles: List<IdentifiedFile>,
+    val markedAsIdentifiedFiles: List<MarkedAsIdentifiedFile>,
+    val listIgnoredFiles: List<IgnoredFile>
+)
+
+/**
+ * A simple Triple bean to hold FossId mapped results
```
Comment (oheger-bosch) [inline scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt:40]: Again usage of "bean", missing dot.

**Review by oheger-bosch — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 3]
```diff
@@ -0,0 +1,381 @@
...
+        val results = mutableMapOf<Package, MutableList<ScanResult>>()
+
+        packages
+            .forEach { pkg: Package ->
+                coroutineScope {
+                    launch {
+                        results.computeIfAbsent(pkg) { mutableListOf() }
+                        val startTime = Instant.now()
+
+                        require(pkg.vcsProcessed.type == VcsType.GIT) {
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:188]: This is a good idea. I will put a TODO in code to remember it.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 4]
```diff
@@ -0,0 +1,381 @@
...
+        outputDirectory: File,
+        downloadDirectory: File
+    ): Map<Package, List<ScanResult>> {
+        val globalStartTime = Instant.now()
+        val results = mutableMapOf<Package, MutableList<ScanResult>>()
+
+        packages
+            .forEach { pkg: Package ->
+                coroutineScope {
+                    launch {
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:184]: You are right. I will remove the coroutineScope.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 5]
```diff
@@ -0,0 +1,381 @@
...
+    }
+
+    /**
+     * Wait until the repository of a scan with [scanCode] has been downloaded.
+     */
+    private suspend fun waitDownloadComplete(scanCode: String) {
+        var done = false
+        var waited = 0
+
+        coroutineScope {
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:264]: Removed.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 6]
```diff
@@ -0,0 +1,381 @@
...
+        return results
+    }
+
+    /**
+     * Create a new scan in the FossID server and return the scan id.
+     */
+    private fun createScan(projectCode: String, url: String, revision: String): String {
+        val formatter = DateTimeFormatter.ofPattern("'$projectCode'_yyyyMMdd_HHmmss")
+        val scanCode = formatter.format(LocalDateTime.now())
+        val response = service.createScan(user, apiKey, projectCode, scanCode, url, revision).execute().body()
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:228]: There is already a function `checkResponse` that aggregates several checks.
I will also check for non-null receiver and hence remove several `requireNotNull(response)` tests.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 7]
```diff
@@ -0,0 +1,381 @@
...
+
+    /**
+     * Wait until the repository of a scan with [scanCode] has been downloaded.
+     */
+    private suspend fun waitDownloadComplete(scanCode: String) {
+        var done = false
+        var waited = 0
+
+        coroutineScope {
+            while (!done && waited < 50) {
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:265]: I rewrote it with `repeat()`. Please review it I am no 100% happy with the result.

[the remark this answers is not in the document]
Comment (oheger-bosch) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: You are right, this does not make much of a difference. I was hoping that Kotlin would already offer something out of the box for waiting repeatedly on a specific event, but maybe not. Given that, I even tend to find the original version a bit easier, or what do you think? In any case, the both wait functions should use the same approach, which is currently not the case.

**Review by oheger-bosch — CHANGES_REQUESTED** [review]
Good improvements, we are getting closer.

**Comment by sschuberth** [comment]
> I was hoping that Kotlin would already offer something out of the box for waiting repeatedly on a specific event, but maybe not.

Could we make use of [channels](https://kotlinlang.org/docs/reference/coroutines/channels.html) here?

**Comment by oheger-bosch** [comment]
> Could we make use of [channels](https://kotlinlang.org/docs/reference/coroutines/channels.html) here?

Mabe using something like [ticker channels](https://kotlinlang.org/docs/reference/coroutines/channels.html#ticker-channels) would be an option. But honestly, I doubt that this would simplify things.

**Comment by sschuberth** [comment]
> Mabe using something like [ticker channels](https://kotlinlang.org/docs/reference/coroutines/channels.html#ticker-channels) would be an option. But honestly, I doubt that this would simplify things.

Do you believe it would result in more code, or in less code that is harder to understand?

**Comment by nnobelis** [comment]
> > Could we make use of [channels](https://kotlinlang.org/docs/reference/coroutines/channels.html) here?
> 
> Mabe using something like [ticker channels](https://kotlinlang.org/docs/reference/coroutines/channels.html#ticker-channels) would be an option. But honestly, I doubt that this would simplify things.

It seems to me the ticker channel is good for waiting a specific amount of time but not for interrupting the wait beforehand.
I will continue experimenting. The old version also does not make me happy because it's a little too C-like :)

**Comment by oheger-bosch** [comment]
> Do you believe it would result in more code, or in less code that is harder to understand?

I fear it would result in more code that is harder to understand ;-)

No, I am no expert in this area, maybe some experiments can shed some light on this.

**Comment by nnobelis** [comment]
@oheger-bosch I pushed a new version with the requested changes.
I found a design a tad more elegant for the wait, by using the `withTimeoutOrNull` function to handle the timeout.
I also consolidated `waitScanComplete` to be on par with `waitDownloadComplete`.

My experiments with the `Ticker` were not satisfying. Additionally, it is annotated as 'Obsolete'.

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt [hunk 8]
```diff
@@ -32,11 +29,13 @@ private const val SCAN_GROUP = "scans"
 private const val PROJECT_GROUP = "projects"
 
 /**
- * This function verifies that a request to FossID server was successful
+ * This function verifies that a request to FossID server was successful.
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/Extensions.kt:32]: Please move plain doc improvements into a separate commit.

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt [hunk 9]
```diff
@@ -32,11 +29,13 @@ private const val SCAN_GROUP = "scans"
...
  * @param operation the label of the operation, for logging
  * @param withDataCheck if true, the payload will be checked for existence
  */
-fun EntityPostResponseBody<*>.checkResponse(operation: String, withDataCheck: Boolean = true) {
+fun <B, T> B?.checkResponse(operation: String, withDataCheck: Boolean = true):
+        B where B : EntityPostResponseBody<T> {
+    requireNotNull(this)
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/Extensions.kt:38]: Instead of requiring `this` to be not `null` here, why not simply defining this to be an extension function on non-nullable `B` instead of `B?`?

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt [hunk 10]
```diff
@@ -45,8 +44,20 @@ fun EntityPostResponseBody<*>.checkResponse(operation: String, withDataCheck: Bo
             "No Payload received for '$operation'. Additional information: $error"
         }
     }
+    return this
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/Extensions.kt:47]: Nit: Blank line before this one.

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt [hunk 11]
```diff
@@ -45,8 +44,20 @@ fun EntityPostResponseBody<*>.checkResponse(operation: String, withDataCheck: Bo
...
 
+/**
+ * This function executes a request to FossID server and verifies it.
+ *
+ * In case of error [msg] is used in logging to describe the request.
+ *
+ * If [withDataCheck] is set, response's payload will be checked for existence
+ */
+fun <B, T> Call<B>.executeAndCheck(msg: String, withDataCheck: Boolean = true):
+        B where B : EntityPostResponseBody<T> =
```
Comment (sschuberth) [inline clients/fossid-webapp/src/main/kotlin/Extensions.kt:58]: Why not write this as `fun <B : EntityPostResponseBody<T>, T>` above instead?

File: clients/fossid-webapp/src/test/kotlin/FossIdClientNewProjectTest.kt [hunk 12]
```diff
@@ -113,6 +113,7 @@ class FossIdClientNewProjectTest : StringSpec({
         service.createScan(
             "", "",
             PROJECT_CODE,
+            SCAN_CODE,
```
Comment (sschuberth) [inline clients/fossid-webapp/src/test/kotlin/FossIdClientNewProjectTest.kt:116]: I general, all modification to files in the `clients` directory should IMO be split out to a preparing commit.

**Review by sschuberth — CHANGES_REQUESTED** [review]
I haven't reviewed the files in the `scanner` directory yet.

**Comment by oheger-bosch** [comment]
> I found a design a tad more elegant for the wait, by using the `withTimeoutOrNull` function to handle the timeout.
> I also consolidated `waitScanComplete` to be on par with `waitDownloadComplete`.
> 

Thanks, I think this looks better. Just as food for thought: Would it be possible to have a generic `wait` function that contains the timeout, the loop, and the delays and is passed a predicate that indicates when the wait condition is fulfilled? Then it could be reused for both waiting for the download and the scan.

**Review by oheger-bosch — DISMISSED** [review]
My points have been addressed.

File: clients/fossid-webapp/src/main/kotlin/Extensions.kt (hunk 9 above)
Comment (nnobelis) [reply clients/fossid-webapp/src/main/kotlin/Extensions.kt:38]: Because this function is there to enforce non-nullability.
If the receiver was non-nullable, I would have to move the check outside on each call, defeating the purpose of this function.

**Comment by nnobelis** [comment]
> Thanks, I think this looks better. Just as food for thought: Would it be possible to have a generic `wait` function that contains the timeout, the loop, and the delays and is passed a predicate that indicates when the wait condition is fulfilled? Then it could be reused for both waiting for the download and the scan.

This is a good idea. I pushed a new version with this function.

**Review by oheger-bosch — DISMISSED** [review]

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 13]
```diff
@@ -0,0 +1,396 @@
...
+                true
+            } else {
+                /*
+        There is a weird bug with the FossId server with version < 20.2: sometimes the download is
+        complete but it stays in state "NOT FINISHED". Therefore we check the output of Git Fetch to
+        find out if the download is indeed done
+                 */
+                val message = response.message
+                if (message == null || !GIT_FETCH_DONE_PATTERN.matcher(message).find()) {
+                    false
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:290]: In general, the above code could be rewritten as

            if (response.data == DownloadStatus.FINISHED) return@wait true

            val message = response.message
            if (message == null || !GIT_FETCH_DONE_PATTERN.matcher(message).matches()) return@wait false

to reduce nesting.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 14]
```diff
@@ -0,0 +1,396 @@
...
+                    }
+                    else -> {
+                        this@FossId.log.info {
+                            "Scan status for '$scanCode' is '${response.data?.state}'. " +
+                                    "Waiting..."
+                        }
+                        false
+                    }
+                }
+            }
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:331]: Probably a bit more readable:

            response.data?.let {
                if (it.state == ScanState.FINISHED) {
                    FossId.log.info { "Scan result: ${response.data?.comment}." }
                    true
                } else {
                    FossId.log.info { "Scan status for '$scanCode' is '${response.data?.state}'. Waiting..." }
                    false
                }
            } ?: false

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 15]
```diff
@@ -0,0 +1,396 @@
...
+        }
+        requireNotNull(timeout) {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:333]: s/timeout/result/

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 16]
```diff
@@ -0,0 +1,396 @@
...
+        val responseIdentifiedFiles = service.listIdentifiedFiles(user, apiKey, scanCode)
+            .checkResponse("list identified files", false)
+
+        // TODO Replace the return type of listIdentifiedFiles to Call<EntityPostResponseBody<Any>> and use toList().
+        val identifiedFiles = responseIdentifiedFiles.data?.values?.toList().orEmpty()
+        log.info("${identifiedFiles.size} identified files have been returned for scan_code=$scanCode.")
+
+        val responseMarkedAsIdentified = service.listMarkedAsIdentifiedFiles(user, apiKey, scanCode)
+            .checkResponse("list marked as identified files", false)
+        requireNotNull(responseMarkedAsIdentified)
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:354]: This is redundant as the type is not nullable.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 17]
```diff
@@ -0,0 +1,396 @@
...
+    private fun createResultSummary(startTime: Instant, rawResults: RawResults): ScanResult {
+        val associate = rawResults.listIgnoredFiles.associateBy { it.path }
+
+        val (filesCount, licenseFindings, copyrightFindings) = kotlin.run {
+            if (rawResults.markedAsIdentifiedFiles.isNotEmpty()) {
+                rawResults.markedAsIdentifiedFiles.mapSummary(associate)
+            } else {
+                rawResults.identifiedFiles.mapSummary(associate)
+            }
+        }
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossId.kt:380]: val (filesCount, licenseFindings, copyrightFindings) = if (rawResults.markedAsIdentifiedFiles.isNotEmpty()) {
            rawResults.markedAsIdentifiedFiles
        } else {
            rawResults.identifiedFiles
        }.mapSummary(associate)

File: scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+    val filesCount: Int,
+    val licenseFindings: MutableList<LicenseFinding>,
+    val copyrightFindings: MutableList<CopyrightFinding>
+)
+
+/**
+ * Map a fossId Raw result to sections that can be included in a [org.ossreviewtoolkit.model.ScanSummary].
+ */
+internal fun <T> List<T>.mapSummary(ignoredFiles: Map<String, IgnoredFile>): FindingsContainer
+        where T : Summarizable {
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt:52]: Please again move the type to `<T : Summarizable>` above.

File: scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+            if (it.isNotEmpty()) {
+                copyrightFindings.add(CopyrightFinding(it, location))
+            }
+        }
+    }
+    return FindingsContainer(
+        filesCount = count,
+        licenseFindings = licenseFindings,
+        copyrightFindings = copyrightFindings
+    )
```
Comment (sschuberth) [inline scanner/src/main/kotlin/scanners/fossid/FossIdScanResults.kt:77]: val files = filterNot { ignoredFiles.contains(it.getFileName()) }

    files.forEach { summarizable ->
        val summary = summarizable.toSummary()
        val location = TextLocation(summary.path, -1, -1)

        summary.licences.forEach {
            val finding = LicenseFinding(it.identifier, location)
            licenseFindings += finding
        }

        summarizable.getCopyright()?.let {
            if (it.isNotEmpty()) {
                copyrightFindings += CopyrightFinding(it, location)
            }
        }
    }

    return FindingsContainer(
        filesCount = files.size,
        licenseFindings = licenseFindings,
        copyrightFindings = copyrightFindings
    )

**Review by sschuberth — CHANGES_REQUESTED** [review]

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 20]
```diff
@@ -0,0 +1,396 @@
...
+        serverUrl = fossIdScannerOptions["serverUrl"] ?: throw IllegalArgumentException("Missing Server Url")
+        apiKey = fossIdScannerOptions["apiKey"] ?: throw IllegalArgumentException("Missing API Key")
+        user = fossIdScannerOptions["user"] ?: throw IllegalArgumentException("Missing User")
+
+        val client = OkHttpClient.Builder().build()
+
+        val mapper = JsonMapper()
+            .registerKotlinModule()
+            .setPropertyNamingStrategy(PropertyNamingStrategies.SNAKE_CASE)
+            .enable(MapperFeature.ACCEPT_CASE_INSENSITIVE_ENUMS)
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:147]: It's needed but I will move it to a future PR.
For now, I will use the mapper from the `FossIdRestService`.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 21]
```diff
@@ -0,0 +1,396 @@
...
+                    delay(delay)
+                }
+            }
+        }
+
+    /**
+     * Wait until the repository of a scan with [scanCode] has been downloaded.
+     */
+    private suspend fun waitDownloadComplete(scanCode: String) {
+        val timeout = wait(WAIT_INTERVAL_MS * WAIT_REPETITION, WAIT_INTERVAL_MS) {
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:274]: IDK. If we change it to have the loop iteration value as first parameter, then the user does not understand what the function does without looking at the implementation.
In the current state, the user see the first param is a timeout, the second is a loopDelay.
With current signature, we can imagine one calling it with a timeout not depending of the loop iteration and I kind of like it this way.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 22]
```diff
@@ -0,0 +1,396 @@
...
+            if (response.data == DownloadStatus.FINISHED) {
+                true
+            } else {
+                /*
+        There is a weird bug with the FossId server with version < 20.2: sometimes the download is
+        complete but it stays in state "NOT FINISHED". Therefore we check the output of Git Fetch to
+        find out if the download is indeed done
+                 */
+                val message = response.message
+                if (message == null || !GIT_FETCH_DONE_PATTERN.matcher(message).find()) {
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:289]: The pattern does not match the entire region therefore I use find()...

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 17 above)
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:380]: Even better, I inverted the test to avoid the negation in the test condition with `isEmpty`.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 23]
```diff
@@ -0,0 +1,396 @@
...
+    private val user: String
+    private val service: FossIdRestService
+
+    // TODO parse the version from the FossID login page
+    override val version = ""
+
+    override val configuration: String
+        get() {
+            // sanitized configuration to hide sensitive information
+            return "--serverUrl=$SECRET --apiKey=$SECRET --user=$SECRET"
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:121]: > More generally, we could discuss whether having a configuration makes sense for remote scanners at all. If we'd never store scan results from remote scanners but always rely on their internal caching, then we'd probably also never need this `configuration` field.

I would not rule out keeping the configuration for remote scanners.
According to companies policies, it can be useful to keep a trace of remote server URLs, API user or even credentials (e.g. if one time token or password was used).

However for FossID, I will simply blank the configuration as suggested.

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt [hunk 24]
```diff
@@ -0,0 +1,396 @@
...
+    override fun getConfigForResult(): ScannerConfiguration {
+        if (config.options == null || config.options?.get("FossId") == null) {
+            return config
+        }
+        val mutableOptions = config.options?.toMutableMap()
+        mutableOptions?.get("FossId")?.toMutableMap()?.let {
+            it.computeIfPresent("serverUrl") { _, _ -> SECRET }
+            it.computeIfPresent("apiKey") { _, _ -> SECRET }
+            it.computeIfPresent("user") { _, _ -> SECRET }
+            mutableOptions.put("FossId", it)
```
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt:157]: See my answer above.

I think that scanner compatibility can also be desired/achieved for remote scanners, by taking in account the security considerations of having the credentials in the config.

**Comment by nnobelis** [comment]
@sschuberth : when you have time, could you please re-review this ?

File: scanner/src/main/kotlin/scanners/fossid/FossId.kt (hunk 20 above)
[the remark this answers is not in the document]
Comment (nnobelis) [reply scanner/src/main/kotlin/scanners/fossid/FossId.kt]: In the time being, the ACCEPT_CASE_INSENSITIVE_ENUMS is already upstream.

**Review by sschuberth — APPROVED** [review]
